capability-providers/keyvalue/rust/src/main.rs

```diff
@@ -0,0 +1,168 @@
+//! # wasmCloud Redis Provider
+//!
+//! Topics relevant to a capability provider:
+//!
+//! RPC:
+//!   * wasmbus.rpc.{prefix}.{provider_key}.{link_name} - Get Invocation, answer InvocationResponse
+//!   * wasmbus.rpc.{prefix}.{public_key}.{link_name}.linkdefs.get - Query all link defs for this provider. (queue subscribed)
+//!   * wasmbus.rpc.{prefix}.{public_key}.{link_name}.linkdefs.del - Remove a link def. Provider de-provisions resources for the given actor.
+//!   * wasmbus.rpc.{prefix}.{public_key}.{link_name}.linkdefs.put - Puts a link def. Provider provisions resources for the given actor.
+//!   * wasmbus.rpc.{prefix}.{public_key}.{link_name}.shutdown - Request for graceful shutdown
+
+#[macro_use]
+extern crate log;
+#[macro_use]
+extern crate lazy_static;
+
+lazy_static! {
+    static ref LINKDEFS: RwLock<HashMap<String, LinkDefinition>> = RwLock::new(HashMap::new());
+}
+
+use crossbeam::sync::Parker;
+use rmp_serde::{Deserializer, Serializer};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::result::Result;
+use std::sync::RwLock;
+
+mod kvredis;
+
+#[derive(Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LinkDefinition {
+    pub actor_id: String,
+    pub provider_id: String,
+    pub link_name: String,
+    pub contract_id: String,
+    pub values: HashMap<String, String>,
+}
+
+#[derive(Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct WasmCloudEntity {
+    pub public_key: String,
+    pub link_name: String,
+    pub contract_id: String,
+}
+
+#[derive(Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Invocation {
+    pub origin: WasmCloudEntity,
+    pub target: WasmCloudEntity,
+    pub operation: String,
+    // may not need this since the host doesn't use it in the Rust invocation
+    //#[serde(with = "serde_bytes")]
+    pub msg: Vec<u8>,
```

Comment:
I think we did figure out we need this from our paired programming this AM, as without it the `msg` parameter will be a list of bytes instead of a byte blob